src/target/metadata.h

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/target/metadata.h
+ * \brief Extends Metadata for use in the compiler.
+ */
+#ifndef TVM_TARGET_METADATA_H_
+#define TVM_TARGET_METADATA_H_
+
+#include <tvm/runtime/metadata.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace tvm {
+namespace target {
+namespace metadata {
+
+class VisitableMetadataNode : public ::tvm::runtime::metadata::MetadataNode {
+ public:
+  explicit VisitableMetadataNode(const struct ::TVMMetadata* data) : MetadataNode{data} {}
+  VisitableMetadataNode() : MetadataNode{nullptr} {}
+
+  void VisitAttrs(AttrVisitor* v) {
+    int64_t version_cpp{version()};
+    v->Visit("version", &version_cpp);
+    auto inputs_array = Array<ObjectRef>();
+    auto inputs_accessor = inputs();
+    inputs_array.reserve(num_inputs());
+    for (int64_t i = 0; i < num_inputs(); ++i) {
+      inputs_array.push_back(::tvm::runtime::metadata::TensorInfo{inputs_accessor[i]});
+    }
+    ::tvm::runtime::metadata::MetadataArray inputs_metadata_array{
+        inputs_array, ::tvm::runtime::metadata::MetadataTypeIndex::kMetadata, "TVMTensorInfo"};
+    v->Visit("inputs", &inputs_metadata_array);
```

Comment:
Should print the size of `inputs` here, otherwise each codegen has to remember that they need to print the size of the array in the array visitor. Just hit this issue in the buggy llvm metadata visitor :)

https://github.com/apache/tvm/blob/4d3d0479da4f4dcf54b05595660c0b31de594a37/include/tvm/runtime/metadata.h#L46-L53

Reply:
the size should be accessible through the array `.size()` right? not sure if you mean something different here?

Reply:
It is, my point is it is better to *explicitly* print the size here inside `src/runtime/target.h` along with other attributes, rather than relying on each visitor not to forget printing after visiting array elements. You have two visitors in the llvm meta data serializer, both of them forgot to print the size so generated types and struct values don't match the expected interface.

Reply:
i see--in offline discussion, @masahi pointed out that other codegen such as LLVM may not have such trivial Visit() functions, and by skipping Visit() on an int64_t, it may increase burden on the Visitor in needing to build a helper function to emit the `num_` fields. additionally, the ordering here is not up to the visitor, so relying on it to maintain the order implicitly in its implementation is a violation of the visitor API. added Visits for num_ to all array fields.